gsy_framework/sim_results/carbon_emissions/results.py

```diff
@@ -0,0 +1,346 @@
+        text = response.text
+
+        if "Unauthorized" in text:
+            raise ValueError("Invalid API key")
+        if "The amount of requested data exceeds allowed limit." in text:
+            raise EntsoeDataLimitExceededError(
+                "The amount of requested data exceeds allowed limit."
+            )
+
+        df_generation = parse_generation(text, per_plant=True, include_eic=False)
+
+        # check if empty
+        if df_generation.empty:
+            raise ValueError("The Entsoe dataframe is empty. Please check the query parameters.")
+        # localize or covert to UTC+0
+        if df_generation.index.tzinfo is not None:
+            df_generation.index = df_generation.index.tz_convert("UTC")
+        else:
+            df_generation.index = df_generation.index.tz_localize("UTC")
+
+        # df_generation is a MultiIndex DataFrame
+        df_generation.columns = df_generation.columns.set_levels(
+            df_generation.columns.levels[0].str.encode("latin-1").str.decode("utf-8"), level=0
+        )
+        # Truncation will fail if data is not sorted along the index
+        df_generation = df_generation.sort_index(axis=0)
+        if df_generation.columns.nlevels == 2:
+            df_generation = (
+                df_generation.assign(newlevel="Actual Aggregated")
+                .set_index("newlevel", append=True)
+                .unstack("newlevel")
+            )
+        df_generation = df_generation.loc[start:end]
+
+        return df_generation
+
+
+class CarbonEmissionsHandler:
+    """The most recent Enstsoe-y version (v0.6.16) is only compatible with python 3.9.
+    Therefore, I am using v0.4.4 whose URL does not work. So, I am overwriting some methods
+    from https://github.com/EnergieID/entsoe-py (branch v0.4.4) to make this work.
+    """
+
+    def __init__(self, entsoe_api_key: str = None):
+        self.entsoe_api_key = entsoe_api_key
+
+    def coordinates_to_country_code(self, lat: float, lon: float) -> str:
+        """Get the country code based on the coordinates"""
+        geolocator = Nominatim(user_agent="location_app")
+        location = geolocator.reverse((lat, lon), language="en")
+        if location and "country" in location.raw["address"]:
+            country_code = location.raw["address"]["country_code"].upper()
+            return country_code
+        return "Country not found"
+
+    def _find_start_and_end_dates(
+        self, dates: List, datetime_format: str = DATE_TIME_FORMAT
+    ) -> Tuple[datetime, datetime]:
+        """Find the start and end dates from a list of string timestamps"""
+
+        dates = [datetime.strptime(timestr, datetime_format) for timestr in dates]
+        start_date = min(dates).replace(tzinfo=pytz.UTC).replace(minute=0, second=0, microsecond=0)
+        end_date = max(dates).replace(tzinfo=pytz.UTC).replace(minute=0, second=0, microsecond=0)
+        return start_date, end_date
```

Comment:
If you always overwriting the `datetime_format` when calling this method, the default value does not make sense.
You are propably overwriting it because it did not work for `strptime`, right?
This is because this format string is for pendulum datetimes.
If you would followed my original suggestion, you would have been more lucky:
https://github.com/gridsingularity/gsy-framework/pull/556#discussion_r1925118474
Let me rewrite this method how I think it should look:
```suggestion
self, times: List) -> Tuple[datetime, datetime]:
        """Find the start and end dates from a list of string timestamps"""

        dates = [from_format(timestr, DATE_TIME_FORMAT) for timestr in times]
        start_date = min(dates).in_tz(TIME_ZONE).set(minute=0, second=0, microsecond=0)
        end_date = max(dates).in_tz(TIME_ZONE).set(minute=0, second=0, microsecond=0)
        return start_date, end_date
```
This will not only enforce our universally used DATE_TIME_FORMAT but also TIME_ZONE.

I have an additional question. Why do we select only full days here and not allow also to query for multiple hours or partial days?